Content Labelling FEP (#116)
swicg/activitypub-trust-and-safety: feps/content-labels/content-labels.md

---
slug: "4c96"
authors: Emelia Smith <emelia@brandedcode.com>
status: DRAFT
# dateReceived: 1970-01-01
# discussionsTo: https://forum.example/topics/xxxx
---

# FEP-4c96: Content Labelling

## Context

> In the open social web, a common issue that is faced by people is being able to control what content they do see, and what content they don't. At present the predominant mechanisms for this in the fediverse is the use of "content warnings", which is where the Object (i.e., Note) is created with a `summary` property and the `as:sensitive` property set to true, however, these content warnings only enable the client to collapse the post.

Reviewer: ```suggestion
In the open social web, a common issue that is faced by people is being able to control what content they see, and what content they don't. At present, the predominant mechanism for this in the fediverse is the use of "content warnings", which is where the Object (i.e., Note) is created with a `summary` property and the `as:sensitive` property set to true; however, these content warnings only enable the client to collapse the post.
```

Author: I think i want to keep the symmetry of "what content they do see, and what content they don't" but perhaps shift the verb:

> what content they do and don't see

Here is an example of what a "content warning" looks like on Mastodon 4.5, when collapsed (initial state):

![Collapsed Content Warning on Mastodon](./images/mastodon-cw-collapsed.png)

And when expanded to reveal the content that was hidden:

![Expanded Content Warning on Mastodon](./images/mastodon-cw-expanded.png)

> The text of the "content warning" can be anything, it doesn't have to be "CW example".

Reviewer: ```suggestion
The text of the "content warning" can be anything; it doesn't have to be "CW example".
```

> The major drawbacks to this style of content filtering is that:

Reviewer: ```suggestion
The major drawbacks to this style of content filtering are the following:
```

> 1. It requires the author of the post to add the content warning when publishing the post. This means that they need to effectively self-censor as their post is subsequentially collapsed by default in their followers' feeds.
> 2. The only mechanism through which for people to filter by "content warning" is by using filters that match for "does this content warning (or post) contain this word". This may also be a regular expression in some cases. This suffers from the scunthorpe problem, and also breaks when someone uses say "k\*\*tens" instead of "kittens" in their content warning text.
> 3. The way content warnings were federated in ActivityPub software was to use the `summary` property (Mastodon, June 18 2017, [PR #3844](https://github.com/mastodon/mastodon/pull/3844)), this now causes problems with blogging and publishing platforms wanting to federate Articles with summaries, since a lot of software will render the sheer existence of `summary` as a content warning.
> 4. The content is always collapsed, unless you opt-out of content warnings all together. Someone cannot currently choose to expand some content warnings but not others. This removes agency of the viewer.
> 5. Related to point one, the content warning can only be applied but the publishing actor, there isn't really any mechanism through which an third-party can apply a content warning to a post.

Reviewer: The order of the items in this list do not seem to be important (except for the reference to the first item by the last); therefore, I'm changing from ordered list to unordered list (and moving the last item to be the second item).
```suggestion
- It requires the author of the post to add the content warning when publishing the post. This means that they need to effectively self-censor as their post is subsequently collapsed by default in their followers' feeds.
- Related to the above, the content warning can only be applied by the publishing actor; there isn't really any mechanism through which a third-party can apply a content warning to a post.
- The only mechanism through which people can filter by "content warning" is to use filters that match for "does this content warning (or post) contain this word?" This may be a regular expression, in some cases. This suffers from the scunthorpe problem, and also breaks when someone uses something like "`k\*\*tens`" instead of "`kittens`" in their content warning text.
- Content warnings were federated in ActivityPub software by use of the `summary` property (Mastodon, June 18 2017, [PR #3844](https://github.com/mastodon/mastodon/pull/3844)). This now causes problems with blogging and publishing platforms wanting to federate Articles with summaries, since a lot of software will render the simple existence of a `summary` value as a content warning.
- The content is always collapsed, unless a user opts out of content warnings altogether. A user cannot currently choose to expand some content warnings but not others. This removes agency of the viewer.
```

Author: I'm fine if you want to swap from numbers to bullets, but mixing in copy changes makes this hard to review & apply

### Rationale

Content Labels exist for two specific purposes: enabling people to filter the content they do and do not wish to see, and enabling service providers to honour age-restriction obligations that apply to them under law. They are not a general-purpose tagging, topic, or organisational system; the `tag` array already supports `Hashtag` and other terms for those purposes.

## Label Definitions

A `LabelDefinition` is the canonical record of a label _term_. It is distinct from a `Label` (the _use_ of a term on a specific object, defined in [Application of Label Definitions](#application-of-label-definitions) below). The split exists so that the application of a label — including who applied it — can carry its own properties without conflating them with the definition of the term itself.

A `LabelDefinition` extends `Object` and is identified by a resolvable IRI. The defining properties are:

| Property                 | Required | Notes                                                                                                                                                           |

> | ------------------------ | -------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |

Reviewer: ```suggestion
| ------------------------ |:--------:| --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
```

Author: This is the correct github formatted markdown as-is

  },
  "url": "https://labels.example/docs/nsfw",
  "context": "https://labels.example/vocab"
}
```

### Behavioural Hints for Labels

> [!NOTE]
> These behavioural hints are subject to change, and the below is a rough draft. The editor of this document will remove this note when the hints reach group consensus.

A `LabelDefinition` MAY also carry additional behavioural-hint properties intended for consuming software:

| Property        | Notes                                                                                                  |
| --------------- | ------------------------------------------------------------------------------------------------------ |
| `severity`      | Suggested severity tier for the label, e.g. `"inform"`, `"alert"`.                                     |
| `blurs`         | What a client SHOULD blur if it chooses to act on the label, e.g. `"content"`, `"media"`, `"nothing"`. |
| `defaultAction` | Suggested default client action, e.g. `"warn"`, `"hide"`, `"ignore"`.                                  |
| `adultOnly`     | Boolean. `true` if the label applies only to adult-restricted content.                                 |

> | `selfLabel`     | Boolean. `true` if the label is intended primarily for first-party application.                        |

Reviewer: ```suggestion
| `severity`      | Suggested severity tier for the label, e.g., `"inform"`, `"alert"`.                                     |
| `blurs`         | What a client SHOULD blur if it chooses to act on the label, e.g., `"content"`, `"media"`, `"nothing"`. |
| `defaultAction` | Suggested default client action, e.g., `"warn"`, `"hide"`, `"ignore"`.                                  |
| `adultOnly`     | Boolean. `true` if the label applies only to adult-restricted content.                                  |
| `selfLabel`     | Boolean. `true` if the label is intended primarily for first-party application.                         |
```

The text of the "content warning" can be anything; it doesn't have to be "CW example".

The major drawbacks to this style of content filtering are the following:

> 1. It requires the author of the post to add the content warning when publishing the post. This means that they need to effectively self-censor as their post is subsequentially collapsed by default in their followers' feeds.
> 2. The only mechanism through which for people to filter by "content warning" is by using filters that match for "does this content warning (or post) contain this word". This may also be a regular expression in some cases. This suffers from the scunthorpe problem, and also breaks when someone uses say "k\*\*tens" instead of "kittens" in their content warning text.

Reviewer: Ordered/numbered lists usually flow from one item to the next, and typically dictate or follow a required sequence. In this case, I'm fairly certain that these two items are independent of one another, and should therefore be in an unordered/bullet list.
```suggestion
- It requires the author of the post to add the content warning when publishing the post. This means that they need to effectively self-censor as their post is subsequently collapsed by default in their followers' feeds.
- The only mechanism through which people can filter on "content warning" is the use of filters that match for something like "does this content warning (or post) contain this word?" This may also be a regular expression in some cases. This, in turn, suffers from the `scunthorpe` problem, and also breaks when someone uses something like `"k\*\*tens"` instead of "`kittens`" in their content warning text.
```
